modules/frontend/tagsharding.go

```diff
@@ -0,0 +1,434 @@
+func (s searchTagSharder) blockMetas(start, end int64, tenantID string) []*backend.BlockMeta {
+	// reduce metas to those in the requested range
+	allMetas := s.reader.BlockMetas(tenantID)
+	metas := make([]*backend.BlockMeta, 0, len(allMetas)/50) // divide by 50 for luck
+	for _, m := range allMetas {
+		if m.StartTime.Unix() <= end &&
+			m.EndTime.Unix() >= start {
+			metas = append(metas, m)
+		}
+	}
+
+	return metas
+}
+
+// backendRequest builds backend requests to search backend blocks. backendRequest takes ownership of reqCh and closes it.
+// it returns 3 int values: totalBlocks, totalBlockBytes, and estimated jobs
+func (s searchTagSharder) backendRequests(ctx context.Context, tenantID string, parent *http.Request, searchReq tagSearchReq, reqCh chan<- *backendReqMsg, stopCh <-chan struct{}) {
+	var blocks []*backend.BlockMeta
+
+	// request without start or end, search only in ingester
+	if searchReq.start() == 0 || searchReq.end() == 0 {
+		close(reqCh)
+		return
+	}
+
+	// calculate duration (start and end) to search the backend blocks
+	start, end := backendRange(searchReq.start(), searchReq.end(), s.cfg.QueryBackendAfter)
+
+	// no need to search backend
+	if start == end {
+		close(reqCh)
+		return
+	}
+
+	// get block metadata of blocks in start, end duration
+	blocks = s.blockMetas(int64(start), int64(end), tenantID)
+
+	targetBytesPerRequest := s.cfg.TargetBytesPerRequest
+
+	go func() {
+		s.buildBackendRequests(ctx, tenantID, parent, blocks, targetBytesPerRequest, reqCh, stopCh, searchReq)
+	}()
+}
+
+// buildBackendRequests returns a slice of requests that cover all blocks in the store
+// that are covered by start/end.
+func (s searchTagSharder) buildBackendRequests(ctx context.Context, tenantID string, parent *http.Request, metas []*backend.BlockMeta, bytesPerRequest int, reqCh chan<- *backendReqMsg, stopCh <-chan struct{}, searchReq tagSearchReq) {
+	defer close(reqCh)
+
+	for _, m := range metas {
+		pages := pagesPerRequest(m, bytesPerRequest)
+		if pages == 0 {
+			continue
+		}
+
+		blockID := m.BlockID.String()
+		for startPage := 0; startPage < int(m.TotalRecords); startPage += pages {
+			subR := parent.Clone(ctx)
+			subR.Header.Set(user.OrgIDHeaderName, tenantID)
+			subR, err := searchReq.buildTagSearchBlockRequest(subR, blockID, startPage, pages, m)
+			if err != nil {
+				reqCh <- &backendReqMsg{err: err}
+				return
+			}
+			subR.RequestURI = buildUpstreamRequestURI(parent.URL.Path, subR.URL.Query())
+			select {
+			case reqCh <- &backendReqMsg{req: subR}:
+			case <-stopCh:
+				return
+			}
+		}
+	}
+}
+
+// ingesterRequest returns a new start and end time range for the backend as well as a http request
+// that covers the ingesters. If nil is returned for the http.Request then there is no ingesters query.
+// since this function modifies searchReq.Start and End we are taking a value instead of a pointer to prevent it from
+// unexpectedly changing the passed searchReq.
```

Comment:
Not sure what you mean, this is actually true for the case I'm using. but I agree because `tagSearchReq` is an interface, this might not be true always.